### PR TITLE
Use real ALIAS target for flann to preserve properties on original target.

### PR DESCRIPTION
**Pull request by 3Descape**
With the current method all properties set on the flann::flann_cpp and flann::flann_cpp target are lost, as they are not copied to the new FLANN::FLANN interface target.
However, there are cases, where we would like preserve these, e.g. to keep the INTERFACE_INCLUDE_DIRECTORIES property that includes the path to the LZ4 headers. Therefore this patch first checks for the appropriate SHARED or STATIC version and then aliases FLANN::FLANN to the respective library.

### CODE REVIEW

File: cmake/Modules/FindFLANN.cmake [hunk 1]
```diff
@@ -48,33 +48,30 @@ if(flann_FOUND)
...
     endif()
   elseif(TARGET flann::flann_cpp_s)
-    set_property(TARGET FLANN::FLANN APPEND PROPERTY INTERFACE_LINK_LIBRARIES flann::flann_cpp_s)
     set(FLANN_LIBRARY_TYPE STATIC)
   else()
-    set_property(TARGET FLANN::FLANN APPEND PROPERTY INTERFACE_LINK_LIBRARIES flann::flann_cpp)
     set(FLANN_LIBRARY_TYPE SHARED)
   endif()
 
+  if(FLANN_LIBRARY_TYPE EQUAL SHARED)
```
Comment (larshg) [inline cmake/Modules/FindFLANN.cmake:69]:
```suggestion
  if(FLANN_LIBRARY_TYPE MATCHES SHARED)
```

**Comment by 3Descape** [comment]
@larshg thank you very much for the suggestion! I appreciate it a lot.

**Comment by mvieth** [comment]
This is relevant for platforms where `flann-config.cmake` is available, right? So for example vcpkg as well as Ubuntu 23.04 and newer (they have Flann 1.9.2)?
Do we have to change PCL to require CMake 3.11 for this ( https://cmake.org/cmake/help/latest/command/add_library.html#alias-libraries )? Currently, the minimum CMake version is 3.10.

**Comment by 3Descape** [comment]
At first a short disclaimer, I am by far not a cmake expert, so please don't take everything with a grain of salt

**"This is relevant for platforms where flann-config.cmake is available, right?"**
Most likely, yes. I personally needed that change, since I installed pcl, flann(1.9.2) and all of the dependencies from source (on windows 10) and without that change MSVC couldn't find the LZ4 headers and include libraries (unless I manually linked against LZ4 in my root project). And since modern cmake already has mechanisms to deal with sub-dependencies in a nice and convenient way, I suggested this fix to take advantage of the information already available on the targets.

**"Do we have to change PCL to require CMake 3.11 for this"**
From what I have seen, no. When looking at the [add_library dokumentation](https://cmake.org/cmake/help/v3.10/command/add_library.html#alias-libraries), we see that cmake 3.10 already supports ALIAS targets. [Cmake 3.11 would be required](https://cmake.org/cmake/help/latest/command/add_library.html#alias-libraries) if we would ALIAS a `IMPORTED GLOBAL` target, but looking at the `flann-targets.cmake` file I have on my machine, I can see that the imported libraries are created as normal imported libraries `add_library(flann::flann_cpp SHARED IMPORTED)`. So no need for Cmake 3.11 from what I understand.

Thank you for reviewing!
If you have any other questions, please let me know :)

**Comment by 3Descape** [comment]
Ah, I see, I missed the "The <target> may not be an Imported Target ". 

I implemented your suggestion to use proper ALIAS targets only for Cmake >= 3.18.0.

Maybe you could check again, if it now works as expected. 
At least for Cmake 3.17.3 and 3.27.6 it seemed to work. 

Thank you very much for the suggestions.

File: cmake/Modules/FindFLANN.cmake [hunk 2]
```diff
@@ -48,33 +48,41 @@ if(flann_FOUND)
...
   else()
-    set_property(TARGET FLANN::FLANN APPEND PROPERTY INTERFACE_LINK_LIBRARIES flann::flann_cpp)
     set(FLANN_LIBRARY_TYPE SHARED)
   endif()
 
+  if(CMAKE_VERSION VERSION_LESS 3.18.0)
+    # Create interface library that effectively becomes an alias for the appropriate (static/dynamic) imported FLANN target
+    add_library(FLANN::FLANN INTERFACE IMPORTED)
+
+    if(FLANN_LIBRARY_TYPE EQUAL SHARED)
```
Comment (mvieth) [inline cmake/Modules/FindFLANN.cmake:73]:
```suggestion
    if(FLANN_LIBRARY_TYPE MATCHES SHARED)
```

File: cmake/Modules/FindFLANN.cmake [hunk 3]
```diff
@@ -48,33 +48,41 @@ if(flann_FOUND)
...
+      set_property(TARGET FLANN::FLANN APPEND PROPERTY INTERFACE_LINK_LIBRARIES flann::flann_cpp)
+    else()
+      set_property(TARGET FLANN::FLANN APPEND PROPERTY INTERFACE_LINK_LIBRARIES flann::flann_cpp_s)
+    endif()
+  else()
+    if(FLANN_LIBRARY_TYPE EQUAL SHARED)
```
Comment (mvieth) [inline cmake/Modules/FindFLANN.cmake:79]:
```suggestion
    if(FLANN_LIBRARY_TYPE MATCHES SHARED)
```

**Review by larshg — APPROVED** [review]

**Pull request merged**